File: android/app/src/main/java/com/phim4k/cinema/StreamResolverPlugin.java
package com.phim4k.cinema;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.util.regex.Pattern;
import javax.net.ssl.HttpsURLConnection;
import org.json.JSONObject;

@CapacitorPlugin(name = "StreamResolver")
public class StreamResolverPlugin extends Plugin {
    private static final Pattern HOST = Pattern.compile("^embed\\d{1,3}\\.streamc\\.xyz$", Pattern.CASE_INSENSITIVE);
    private static final Pattern HASH = Pattern.compile("^[a-f0-9]{32}$", Pattern.CASE_INSENSITIVE);

    @PluginMethod
    public void resolve(PluginCall call) {
        new Thread(() -> {
            try {
                URI target = URI.create(call.getString("url", ""));
                URI referrer = URI.create(call.getString("referrer", ""));
                String query = target.getRawQuery() == null ? "" : target.getRawQuery();
                String hash = query.startsWith("hash=") && query.indexOf('&') < 0 ? query.substring(5) : "";
                if (!"https".equalsIgnoreCase(target.getScheme()) || !HOST.matcher(target.getHost()).matches()
                    || !"/embed.php".equals(target.getPath()) || !HASH.matcher(hash).matches()
                    || !"https".equalsIgnoreCase(referrer.getScheme()) || !"phim.nguonc.com".equalsIgnoreCase(referrer.getHost())) {
                    throw new IllegalArgumentException("Invalid backup stream request.");
                }
                HttpsURLConnection connection = (HttpsURLConnection) target.toURL().openConnection();
                connection.setRequestMethod("POST");
                connection.setConnectTimeout(15000);
                connection.setReadTimeout(30000);
                connection.setDoOutput(true);
                connection.setRequestProperty("Accept", "application/json, text/plain, */*");
                connection.setRequestProperty("Accept-Language", "vi,en-US;q=0.8,en;q=0.6");
                connection.setRequestProperty("Content-Type", "application/json");
                connection.setRequestProperty("Origin", target.getScheme() + "://" + target.getHost());
                connection.setRequestProperty("Referer", target.toString());
                connection.setRequestProperty("User-Agent", "Mozilla/5.0 (Linux; Android 14) AppleWebKit/537.36 Mobile Safari/537.36 Phim4KAndroid");
                JSONObject body = new JSONObject()
                    .put("action", "bootstrap").put("referrer", referrer.toString())
                    .put("frame_origins", new org.json.JSONArray().put("https://phim.nguonc.com"))
                    .put("request_grant", true).put("playlist_format", "hls")
                    .put("pretty_url", true).put("path_chunks", true).put("bootstrap_format", "json");
                byte[] bytes = body.toString().getBytes(StandardCharsets.UTF_8);
                connection.setFixedLengthStreamingMode(bytes.length);
                try (OutputStream output = connection.getOutputStream()) { output.write(bytes); }
                if (connection.getResponseCode() != 200) throw new IllegalStateException("Backup stream rejected (" + connection.getResponseCode() + ").");
                StringBuilder text = new StringBuilder();
                try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
                    for (String line; (line = reader.readLine()) != null;) text.append(line);
                }
                JSONObject preissued = new JSONObject(text.toString()).getJSONObject("preissued");
                URI playlist = URI.create(preissued.getString("playlist"));
                if (!"hls".equals(preissued.optString("playlistFormat")) || !"https".equalsIgnoreCase(playlist.getScheme())
                    || !target.getHost().equalsIgnoreCase(playlist.getHost())) throw new IllegalStateException("Invalid backup playlist.");
                call.resolve(new JSObject().put("playlist", playlist.toString()));
            } catch (Exception error) {
                call.reject(error.getMessage() == null ? "Backup stream failed." : error.getMessage());
            }
        }, "phim4k-stream-resolver").start();
    }
}

File: ios/App/App/StreamResolverPlugin.swift
import Capacitor
import Foundation

@objc(StreamResolverPlugin)
public final class StreamResolverPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "StreamResolverPlugin"
    public let jsName = "StreamResolver"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "resolve", returnType: CAPPluginReturnPromise),
    ]

    @objc public func resolve(_ call: CAPPluginCall) {
        guard let rawURL = call.getString("url"), let target = URL(string: rawURL),
              target.scheme == "https", target.path == "/embed.php",
              let host = target.host, host.range(of: #"^embed\d{1,3}\.streamc\.xyz$"#, options: [.regularExpression, .caseInsensitive]) != nil,
              let components = URLComponents(url: target, resolvingAgainstBaseURL: false), components.queryItems?.count == 1,
              components.queryItems?.first?.name == "hash",
              let hash = components.queryItems?.first?.value,
              hash.range(of: #"^[a-f0-9]{32}$"#, options: [.regularExpression, .caseInsensitive]) != nil,
              let rawReferrer = call.getString("referrer"), let referrer = URL(string: rawReferrer),
              referrer.scheme == "https", referrer.host == "phim.nguonc.com" else {
            call.reject("Invalid backup stream request.")
            return
        }
        var request = URLRequest(url: target, timeoutInterval: 30)
        request.httpMethod = "POST"
        request.setValue("application/json, text/plain, */*", forHTTPHeaderField: "Accept")
        request.setValue("vi,en-US;q=0.8,en;q=0.6", forHTTPHeaderField: "Accept-Language")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("https://\(host)", forHTTPHeaderField: "Origin")
        request.setValue(target.absoluteString, forHTTPHeaderField: "Referer")
        request.setValue("Mozilla/5.0 (iPhone; CPU iPhone OS 18_0 like Mac OS X) AppleWebKit/605.1.15 Mobile/15E148 Phim4KiOS", forHTTPHeaderField: "User-Agent")
        request.httpBody = try? JSONSerialization.data(withJSONObject: [
            "action": "bootstrap", "referrer": referrer.absoluteString,
            "frame_origins": ["https://phim.nguonc.com"], "request_grant": true,
            "playlist_format": "hls", "pretty_url": true, "path_chunks": true, "bootstrap_format": "json",
        ])
        URLSession.shared.dataTask(with: request) { data, response, error in
            guard error == nil, let http = response as? HTTPURLResponse, http.statusCode == 200, let data,
                  let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let preissued = object["preissued"] as? [String: Any], preissued["playlistFormat"] as? String == "hls",
                  let rawPlaylist = preissued["playlist"] as? String, let playlist = URL(string: rawPlaylist),
                  playlist.scheme == "https", playlist.host?.caseInsensitiveCompare(host) == .orderedSame else {
                call.reject("Backup stream could not be resolved.")
                return
            }
            call.resolve(["playlist": playlist.absoluteString])
        }.resume()
    }
}
